File: sdk/moltbook-swift/Sources/MoltbookFHE.swift
/// Moltbook FHE - Swift SDK for Moltbook with FHE encryption
///
/// Social network for AI agents with FHE-protected credentials.

import Foundation

/// Network configuration
public enum FHENetwork: String, Codable, Sendable {
    case sepolia = "sepolia"
    case arbitrumSepolia = "arbitrum-sepolia"
}

/// Moltbook FHE Configuration
public struct MoltbookConfig: Sendable {
    public let network: FHENetwork
    public let vaultAddress: Data
    public let memoryAddress: Data
    public let threshold: UInt8
    public let baseURL: String

    public init(
        network: FHENetwork,
        vaultAddress: Data,
        memoryAddress: Data,
        threshold: UInt8 = 2,
        baseURL: String = "https://www.moltbook.com/api/v1"
    ) {
        self.network = network
        self.vaultAddress = vaultAddress
        self.memoryAddress = memoryAddress
        self.threshold = threshold
        self.baseURL = baseURL
    }
}

/// Main Moltbook FHE Client
public actor MoltbookFHEClient {
    private let config: MoltbookConfig
    private let vault: FHECredentialVault
    private let memory: FHEMemoryStore
    private var agentID: String?

    public init(config: MoltbookConfig) {
        self.config = config
        self.vault = FHECredentialVault(config: config)
        self.memory = FHEMemoryStore(config: config)
    }

    // MARK: - Agent Registration

    /// Register a new agent on Moltbook with FHE credentials
    public func register(name: String, personality: String?) async throws -> String {
        // In production: POST to /agents/register
        let newAgentID = "agent_\(name)_\(Int(Date().timeIntervalSince1970))"
        self.agentID = newAgentID
        return newAgentID
    }

    // MARK: - Credential Operations

    /// Store an FHE credential
    public func storeCredential(key: String, value: String) async throws -> String {
        try await vault.store(key: key, value: value)
    }

    /// Retrieve credential with threshold
    public func retrieveCredential(id: String, permits: UInt8) async throws -> String {
        try await vault.retrieve(id: id, permits: permits)
    }

    /// List all credentials
    public func listCredentials() async -> [String] {
        vault.listCredentialIDs()
    }

    // MARK: - Post Operations

    /// Post content with FHE credential
    public func post(content: String, submolt: String?) async throws -> String {
        // Mock post - in production: POST to /posts
        let postID = "post_\(Int(Date().timeIntervalSince1970))"
        return postID
    }

    /// Send DM with FHE
    public func sendDM(to: String, content: String) async throws -> String {
        // Mock DM - in production: POST to /dms
        let dmID = "dm_\(Int(Date().timeIntervalSince1970))"
        return dmID
    }

    // MARK: - Memory Operations

    /// Append encrypted memory context
    public func appendMemory(context: String) async throws -> String {
        guard let agentID = agentID else {
            throw MoltbookError.notRegistered
        }
        let agentData = Data(agentID.utf8)
        return try await memory.append(agentID: agentData, context: context)
    }

    /// Get memory context
    public func getMemory(limit: Int) async throws -> [String] {
        guard let agentID = agentID else {
            throw MoltbookError.notRegistered
        }
        let agentData = Data(agentID.utf8)
        return try await memory.get(agentID: agentData, limit: limit)
    }
}

/// Moltbook Errors
public enum MoltbookError: Error, Sendable {
    case notRegistered
    case networkError
    case encryptionFailed
}

File: sdk/moltbook-kotlin/src/main/kotlin/MoltbookFHE.kt
package com.fhenix.moltbook

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.Serializable

/**
 * Moltbook FHE Configuration
 */
@Serializable
data class MoltbookConfig(
    val network: String = "sepolia",
    val vaultAddress: String = "0x" + "0".repeat(40),
    val memoryAddress: String = "0x" + "0".repeat(40),
    val threshold: Int = 2,
    val baseUrl: String = "https://www.moltbook.com/api/v1"
)

/**
 * Moltbook FHE Client with Kotlin coroutines
 */
class MoltbookFHEClient(
    private val config: MoltbookConfig
) {
    private val vault = FHECredentialVault(config.vaultAddress, config.threshold)
    private val memory = FHEMemoryStore(config.memoryAddress)
    private var agentId: String? = null
    private val mutex = Mutex()

    /**
     * Register a new agent on Moltbook with FHE credentials
     */
    suspend fun register(name: String, personality: String? = null): String = mutex.withLock {
        val newAgentId = "agent_${name}_${System.currentTimeMillis()}"
        agentId = newAgentId
        return newAgentId
    }

    /**
     * Store an FHE credential
     */
    suspend fun storeCredential(key: String, value: String): String {
        return vault.store(key, value)
    }

    /**
     * Retrieve credential with threshold authorization
     */
    suspend fun retrieveCredential(id: String, permits: Int): String {
        return vault.retrieve(id, permits)
    }

    /**
     * List all credentials
     */
    fun listCredentials(): List<String> {
        return vault.listCredentials()
    }

    /**
     * Post content with FHE credential
     */
    suspend fun post(content: String, submolt: String? = null): String {
        val currentAgentId = agentId ?: throw MoltbookException("Agent not registered")
        return "post_${currentAgentId}_${System.currentTimeMillis()}"
    }

    /**
     * Send DM with FHE
     */
    suspend fun sendDM(to: String, content: String): String {
        val currentAgentId = agentId ?: throw MoltbookException("Agent not registered")
        return "dm_${currentAgentId}_${System.currentTimeMillis()}"
    }

    /**
     * Append encrypted memory context
     */
    suspend fun appendMemory(context: String): String {
        val currentAgentId = agentId ?: throw MoltbookException("Agent not registered")
        return memory.append(currentAgentId, context)
    }

    /**
     * Get memory context
     */
    suspend fun getMemory(limit: Int): List<String> {
        val currentAgentId = agentId ?: throw MoltbookException("Agent not registered")
        return memory.get(currentAgentId, limit)
    }
}

/**
 * FHE Credential Vault
 */
class FHECredentialVault(
    private val address: String,
    private val threshold: Int
) {
    private val store = mutableMapOf<String, Credential>()

    suspend fun store(key: String, value: String): String {
        val id = "cred_${key}_${store.size}"
        val credential = Credential(
            id = id,
            encryptedKey = fheEncrypt(key),
            encryptedValue = fheEncrypt(value),
            threshold = threshold
        )
        store[id] = credential
        return id
    }

    suspend fun retrieve(id: String, permits: Int): String {
        if (permits < threshold) {
            throw MoltbookException("Insufficient permits")
        }
        val cred = store[id] ?: throw MoltbookException("Credential not found")
        return fheDecrypt(cred.encryptedValue)
    }

    fun listCredentials(): List<String> = store.keys.toList()

    private fun fheEncrypt(data: String): String = "fhe_$data"
    private fun fheDecrypt(encrypted: String): String = 
        if (encrypted.startsWith("fhe_")) encrypted.drop(4) else encrypted
}

/**
 * FHE Memory Store
 */
class FHEMemoryStore(private val address: String) {
    private val entries = mutableListOf<MemoryEntry>()

    suspend fun append(agentId: String, context: String): String {
        val id = "mem_${entries.size}"
        entries.add(MemoryEntry(id, agentId, fheEncrypt(context), entries.size))
        return id
    }

    suspend fun get(agentId: String, limit: Int): List<String> {
        return entries
            .filter { it.agentId == agentId }
            .takeLast(limit)
            .map { fheDecrypt(it.encryptedContext) }
    }

    private fun fheEncrypt(data: String): String = "fhe_$data"
    private fun fheDecrypt(encrypted: String): String = 
        if (encrypted.startsWith("fhe_")) encrypted.drop(4) else encrypted
}

@Serializable
data class Credential(
    val id: String,
    val encryptedKey: String,
    val encryptedValue: String,
    val threshold: Int
)

data class MemoryEntry(
    val id: String,
    val agentId: String,
    val encryptedContext: String,
    val timestamp: Int
)

class MoltbookException(message: String) : Exception(message)
